csharp/ql/src/experimental/CWE-095/razor_parse.ql

```diff
@@ -0,0 +1,77 @@
+/**
+ * @name Server-Side Template Injection in RazorEngine
+ * @description User-controlled data may be evaluated, leading to arbitrary code execution.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id csharp/razor-injection
+ * @tags security
+ */
+
+import csharp
+import semmle.code.csharp.frameworks.microsoft.AspNetCore
+import semmle.code.csharp.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/*
+ * The offending method is Parse from RazorEngine.Razor.
+ */
+
+class RazorEngineClass extends Class {
+  RazorEngineClass() { this.hasQualifiedName("RazorEngine.Razor") }
+
+  Method getParseMethod() {
+    result.getDeclaringType() = this and
+    result.hasName("Parse")
+  }
+}
+
+/*
+ * We are only interested in ASP.NET MVC Controller or ControllerBase classes
+ */
+
+class ControllerMVC extends Class {
+  ControllerMVC() {
+    this.hasQualifiedName("System.Web.Mvc", "Controller") or
+    this.hasQualifiedName("Microsoft.AspNetCore.Mvc", "Controller") or
+    this instanceof MicrosoftAspNetCoreMvcController or
+    this instanceof MicrosoftAspNetCoreMvcControllerBaseClass
+  }
+}
```

Comment:
I would replace this with
```ql
import semmle.code.csharp.frameworks.system.web.Mvc as Mvc
import semmle.code.csharp.frameworks.microsoft.AspNetCore

...

class Controller extends Class {
  Controller() {
    this instanceof Mvc::Controller
    or
    this instanceof MicrosoftAspNetCoreMvcController
  }

  Method getAPostActionMethod() {
    result = this.(Mvc::Controller).getAPostActionMethod()
    or
    result = this.(MicrosoftAspNetCoreMvcController).getAnActionMethod() and
    result.getAnAttribute() instanceof MicrosoftAspNetCoreMvcHttpPostAttribute
  }
}
```

and then `isSource` can be defined as

```ql
source.asParameter() = any(Controller c).getAPostActionMethod().getAParameter()
```

Reply:
Thanks a lot for the feedback. The query does detect the vulnerability in RazorVulnerableApp. You are correct, we could improve this by adding cookie values and URL parameters. It was my first attempt to model the vulnerability, I thought GET/POST requests were covered.  

I will update the code as suggested.

Reply:
Just noticed the class AspNetCoreActionMethodParameter but I don't think I'm using it correctly as I'm not getting the expected results.